### DISCUSSION TITLE
Byte Buddy for real-time systems

### DISCUSSION THREAD

**Discussion by druzhynin-oleksii** [raphw/byte-buddy#1385] [Ideas]
Hi.

We have very special case which exists only in real-time bidding and self-driving.

We have real-time Java agent which has 500+ input parameters from data sources and should answer within 300ms. To debug we need a sample of that input parameters and a outside environment reaction which very hard to simulate. 

So we need to run our engine twice for debug:
1. to get the environment reaction
2. to collect debugged parameters (which set can be varied)

Can we use Byte Buddy for the second purpose? 

I checked https://www.baeldung.com/lombok-custom-annotation to duplicate all the classes to cover both points in real time, but Java code generation looks really as a headache comparing to byte code updates.   

Regards,

Oleksii

**Comment by raphw** [comment]
Yes, you would likely want to use `AgentBuilder` and a Java agent for this. That way, you can rewrite your code on the fly and collect input to methods.

**Comment by druzhynin-oleksii** [comment]
Like here: https://gist.github.com/raphw/34c0e2fffe2ee7b4f02f ?

**Comment by raphw** [comment]
That's an example of the API, yes.

**Comment by druzhynin-oleksii** [comment]
Hi Raphael.

I am just wondering why there is no good library for code modification before compilation. lombok can add only methods, but I didn't find how to make a syntactic sub-class using this.

Alex

**Comment by raphw** [comment]
There is JavaPoet which I personally like, but source code is way less structured what makes it much harder. Also, class file manipulation is more universal as it can target all JVM languages.

**Comment by druzhynin-oleksii** [comment]
But JavaPoet doesn't allow to modify existing classes.
